[WTF] Add more LIFETIME_BOUND attributes to iterator, span and inner pointer methods

ddkilzer opened the pull request:
#### 4f842d8d9a8e1f9610635c931ff9ac31c36ca186
<pre>
[WTF] Add more LIFETIME_BOUND attributes to iterator, span and inner pointer methods
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=285113">https://bugs.webkit.org/show_bug.cgi?id=285113</a>&gt;
&lt;<a href="https://rdar.apple.com/141961625">rdar://141961625</a>&gt;

Reviewed by Geoffrey Garen.

Add more LIFETIME_BOUND attributes to WTF classes unless otherwise noted
below.  Also fix the LIFETIME_BOUND attribute for gcc and MSVC++.

* Source/JavaScriptCore/runtime/ExceptionHelpers.cpp:
(JSC::clampErrorMessage):

* Source/WTF/wtf/Compiler.h:
(LIFETIME_BOUND):
- Fix attribute for gcc, and add attribute for MSVC++.

* Source/WTF/wtf/FileSystem.h:
(WTF::MappedFileData::span):
(WTF::MappedFileData::mutableSpan):
* Source/WTF/wtf/FixedVector.h:
(WTF::FixedVector::begin):
(WTF::FixedVector::end):
(WTF::FixedVector::begin const):
(WTF::FixedVector::end const):
(WTF::FixedVector::rbegin):
(WTF::FixedVector::rend):
(WTF::FixedVector::rbegin const):
(WTF::FixedVector::rend const):
(WTF::FixedVector::at):
(WTF::FixedVector::at const):
(WTF::FixedVector::operator[]):
(WTF::FixedVector::operator[] const):
(WTF::FixedVector::first):
(WTF::FixedVector::first const):
(WTF::FixedVector::last):
(WTF::FixedVector::last const):
(WTF::FixedVector::storage):
(WTF::FixedVector::span const):
(WTF::FixedVector::mutableSpan):
(WTF::FixedVector::subspan const):
* Source/WTF/wtf/MallocSpan.h:
(WTF::MallocSpan::span const):
(WTF::MallocSpan::mutableSpan):
(WTF::MallocSpan::operator[]):
(WTF::MallocSpan::operator[] const):
(WTF::MallocSpan::operator bool const):
- Add !! to match other pointer -&gt; bool conversions.
* Source/WTF/wtf/StackShot.h:
(WTF::StackShot::array const):
(WTF::StackShot::span const):
* Source/WTF/wtf/TrailingArray.h:
(WTF::TrailingArray::data):
(WTF::TrailingArray::data const):
(WTF::TrailingArray::span):
(WTF::TrailingArray::span const):
(WTF::TrailingArray::begin):
(WTF::TrailingArray::end):
(WTF::TrailingArray::begin const):
(WTF::TrailingArray::end const):
(WTF::TrailingArray::cbegin const):
(WTF::TrailingArray::cend const):
(WTF::TrailingArray::rbegin):
(WTF::TrailingArray::rend):
(WTF::TrailingArray::rbegin const):
(WTF::TrailingArray::rend const):
(WTF::TrailingArray::crbegin const):
(WTF::TrailingArray::crend const):
(WTF::TrailingArray::at):
(WTF::TrailingArray::at const):
(WTF::TrailingArray::operator[]):
(WTF::TrailingArray::operator[] const):
(WTF::TrailingArray::first):
(WTF::TrailingArray::first const):
(WTF::TrailingArray::last):
(WTF::TrailingArray::last const):
* Source/WTF/wtf/URL.cpp:
(WTF::substringIgnoringQueryAndFragments):
* Source/WTF/wtf/UUID.h:
(WTF::UUID::span const):
* Source/WTF/wtf/persistence/PersistentEncoder.h:
(WTF::Persistence::Encoder::buffer const):
(WTF::Persistence::Encoder::span const):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::span8 const):
(WTF::StringBuilder::span16 const):
(WTF::StringBuilder::span const):
* Source/WTF/wtf/text/StringConcatenateNumbers.h:
(WTF::StringTypeAdapter::span const):
(WTF::FormattedNumber::span const):
(WTF::FormattedNumber::buffer const):
(WTF::FormattedNumber::span const):
(WTF::FormattedCSSNumber::buffer const):
(WTF::FormattedCSSNumber::span const):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::span8 const):
(WTF::StringImpl::span16 const):
(WTF::StringImpl::span const):
* Source/WTF/wtf/text/StringParsingBuffer.h:
(WTF::StringParsingBuffer::StringParsingBuffer):
(WTF::StringParsingBuffer::position const):
(WTF::StringParsingBuffer::end const):
(WTF::StringParsingBuffer::stringViewOfCharactersRemaining const):
(WTF::StringParsingBuffer::span const):
(WTF::StringParsingBuffer::consume):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::StringView):
(WTF::StringView::rawCharacters const):
(WTF::StringView::span8 const):
(WTF::StringView::span16 const):
(WTF::StringView::span const):
(WTF::StringView::UpconvertedCharactersWithSize::operator const UChar* const):
(WTF::StringView::UpconvertedCharactersWithSize::get const):
(WTF::StringView::UpconvertedCharactersWithSize::operator std::span&lt;const UChar&gt; const):
(WTF::StringView::UpconvertedCharactersWithSize::span const):
(WTF::StringView::SplitResult::SplitResult):
(WTF::StringView::SplitResult::begin const):
(WTF::StringView::SplitResult::end const):
(WTF::StringView::GraphemeClusters::GraphemeClusters):
(WTF::StringView::GraphemeClusters::begin const):
(WTF::StringView::GraphemeClusters::end const):
(WTF::StringView::CodePoints::CodePoints):
(WTF::StringView::CodePoints::begin const):
(WTF::StringView::CodePoints::end const):
(WTF::StringView::CodePoints::codePointAt const):
(WTF::StringView::CodeUnits::CodeUnits):
(WTF::StringView::CodeUnits::begin const):
(WTF::StringView::CodeUnits::end const):
(WTF::StringView::GraphemeClusters::Iterator::Iterator):
(WTF::StringView::CodePoints::Iterator::Iterator):
(WTF::StringView::CodeUnits::Iterator::Iterator):

* Source/WebCore/rendering/RenderListMarker.h:
(WebCore::ListMarkerTextContent::textWithoutSuffix const):
(WebCore::ListMarkerTextContent::suffix const):

Canonical link: <a href="https://commits.webkit.org/288278@main">https://commits.webkit.org/288278@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b5e965c9a2b9f17bf99e06a525cd9a8c6f1cd99

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82288 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1952 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36356 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87420 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33349 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84393 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2023 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9834 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64134 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21882 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85357 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1412 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74875 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44412 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1314 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29056 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32390 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75274 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72607 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29680 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88776 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81342 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9594 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6853 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72526 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9820 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70689 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71744 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15880 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14887 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/949 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12780 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9547 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15068 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103755 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9421 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25171 "Found 3 new JSC stress test failures: stress/json-stringify-inspector-check.js.no-llint, stress/substring-global-atom-cache.js.lockdown, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12887 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11191 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren reviewed (approved):
Longer term we’re going to need to distinguish two potential meanings of LIFETIMEBOUND:

1. the lifetime of a is an upper bound on the lifetime of b (weak guarantee)
2. the lifetime of a is a lower bound on the lifetime of b (strong guarantee) 

In cases where a holds a mutable pointer to b, we can only offer 1. 

I wonder if we should set the precedent now, and have a MUTABLE_LIFETIMEBOUND or WEAK_LIFETIMEBOUND alias for LIFETIMEBOUND (even though we don’t have a way to enforce the difference yet), to put the right framework in place.

Commit:
[WTF] Add more LIFETIME_BOUND attributes to iterator, span and inner pointer methods
<https://bugs.webkit.org/show_bug.cgi?id=285113>
<rdar://141961625>

Reviewed by Geoffrey Garen.

Add more LIFETIME_BOUND attributes to WTF classes unless otherwise noted
below.  Also fix the LIFETIME_BOUND attribute for gcc and MSVC++.

* Source/JavaScriptCore/runtime/ExceptionHelpers.cpp:
(JSC::clampErrorMessage):

* Source/WTF/wtf/Compiler.h:
(LIFETIME_BOUND):
- Fix attribute for gcc, and add attribute for MSVC++.

* Source/WTF/wtf/FileSystem.h:
(WTF::MappedFileData::span):
(WTF::MappedFileData::mutableSpan):
* Source/WTF/wtf/FixedVector.h:
(WTF::FixedVector::begin):
(WTF::FixedVector::end):
(WTF::FixedVector::begin const):
(WTF::FixedVector::end const):
(WTF::FixedVector::rbegin):
(WTF::FixedVector::rend):
(WTF::FixedVector::rbegin const):
(WTF::FixedVector::rend const):
(WTF::FixedVector::at):
(WTF::FixedVector::at const):
(WTF::FixedVector::operator[]):
(WTF::FixedVector::operator[] const):
(WTF::FixedVector::first):
(WTF::FixedVector::first const):
(WTF::FixedVector::last):
(WTF::FixedVector::last const):
(WTF::FixedVector::storage):
(WTF::FixedVector::span const):
(WTF::FixedVector::mutableSpan):
(WTF::FixedVector::subspan const):
* Source/WTF/wtf/MallocSpan.h:
(WTF::MallocSpan::span const):
(WTF::MallocSpan::mutableSpan):
(WTF::MallocSpan::operator[]):
(WTF::MallocSpan::operator[] const):
(WTF::MallocSpan::operator bool const):
- Add !! to match other pointer -> bool conversions.
* Source/WTF/wtf/StackShot.h:
(WTF::StackShot::array const):
(WTF::StackShot::span const):
* Source/WTF/wtf/TrailingArray.h:
(WTF::TrailingArray::data):
(WTF::TrailingArray::data const):
(WTF::TrailingArray::span):
(WTF::TrailingArray::span const):
(WTF::TrailingArray::begin):
(WTF::TrailingArray::end):
(WTF::TrailingArray::begin const):
(WTF::TrailingArray::end const):
(WTF::TrailingArray::cbegin const):
(WTF::TrailingArray::cend const):
(WTF::TrailingArray::rbegin):
(WTF::TrailingArray::rend):
(WTF::TrailingArray::rbegin const):
(WTF::TrailingArray::rend const):
(WTF::TrailingArray::crbegin const):
(WTF::TrailingArray::crend const):
(WTF::TrailingArray::at):
(WTF::TrailingArray::at const):
(WTF::TrailingArray::operator[]):
(WTF::TrailingArray::operator[] const):
(WTF::TrailingArray::first):
(WTF::TrailingArray::first const):
(WTF::TrailingArray::last):
(WTF::TrailingArray::last const):
* Source/WTF/wtf/URL.cpp:
(WTF::substringIgnoringQueryAndFragments):
* Source/WTF/wtf/UUID.h:
(WTF::UUID::span const):
* Source/WTF/wtf/persistence/PersistentEncoder.h:
(WTF::Persistence::Encoder::buffer const):
(WTF::Persistence::Encoder::span const):
* Source/WTF/wtf/text/StringBuilder.h:
(WTF::StringBuilder::span8 const):
(WTF::StringBuilder::span16 const):
(WTF::StringBuilder::span const):
* Source/WTF/wtf/text/StringConcatenateNumbers.h:
(WTF::StringTypeAdapter::span const):
(WTF::FormattedNumber::span const):
(WTF::FormattedNumber::buffer const):
(WTF::FormattedNumber::span const):
(WTF::FormattedCSSNumber::buffer const):
(WTF::FormattedCSSNumber::span const):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::span8 const):
(WTF::StringImpl::span16 const):
(WTF::StringImpl::span const):
* Source/WTF/wtf/text/StringParsingBuffer.h:
(WTF::StringParsingBuffer::StringParsingBuffer):
(WTF::StringParsingBuffer::position const):
(WTF::StringParsingBuffer::end const):
(WTF::StringParsingBuffer::stringViewOfCharactersRemaining const):
(WTF::StringParsingBuffer::span const):
(WTF::StringParsingBuffer::consume):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::StringView):
(WTF::StringView::rawCharacters const):
(WTF::StringView::span8 const):
(WTF::StringView::span16 const):
(WTF::StringView::span const):
(WTF::StringView::UpconvertedCharactersWithSize::operator const UChar* const):
(WTF::StringView::UpconvertedCharactersWithSize::get const):
(WTF::StringView::UpconvertedCharactersWithSize::operator std::span<const UChar> const):
(WTF::StringView::UpconvertedCharactersWithSize::span const):
(WTF::StringView::SplitResult::SplitResult):
(WTF::StringView::SplitResult::begin const):
(WTF::StringView::SplitResult::end const):
(WTF::StringView::GraphemeClusters::GraphemeClusters):
(WTF::StringView::GraphemeClusters::begin const):
(WTF::StringView::GraphemeClusters::end const):
(WTF::StringView::CodePoints::CodePoints):
(WTF::StringView::CodePoints::begin const):
(WTF::StringView::CodePoints::end const):
(WTF::StringView::CodePoints::codePointAt const):
(WTF::StringView::CodeUnits::CodeUnits):
(WTF::StringView::CodeUnits::begin const):
(WTF::StringView::CodeUnits::end const):
(WTF::StringView::GraphemeClusters::Iterator::Iterator):
(WTF::StringView::CodePoints::Iterator::Iterator):
(WTF::StringView::CodeUnits::Iterator::Iterator):

* Source/WebCore/rendering/RenderListMarker.h:
(WebCore::ListMarkerTextContent::textWithoutSuffix const):
(WebCore::ListMarkerTextContent::suffix const):

Canonical link: https://commits.webkit.org/288278@main
